Feature/UI-statemachine

Wildeaso1 opened the pull request.

Wildeaso1 committed:
Created states for UI statemachine

Wildeaso1 committed:
Created StateManager for UI

Wildeaso1 committed:
Created Custom Event/Caller

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateEvent.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace UI.Canvas.StateMachine
+{
+    [System.Serializable]
+    public class UIStateEvent : UnityEvent<StateChange>
+    {
+    }
```
```suggestion
    [System.Serializable]
    public sealed class UIStateEvent : UnityEvent<StateChange> { }
```

Ik werdt erg verward toen ik dit voor het eerst zag, daarom een summary toevoegen is wel handig. [Voorbeeld](https://github.com/Team-Swamp/IceBites/blob/develop/Assets/Scripts/Framework/Cooking/HeldItem.cs)!

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateCaller.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace UI.Canvas.StateMachine
+{
+    public class UIStateCaller : MonoBehaviour
```
```suggestion
    public sealed class UIStateCaller : MonoBehaviour
```

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateCaller.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class UIStateCaller : MonoBehaviour
+    {
+        private UIStateEvent onStateChange;
+        [SerializeField] private UIStateManager stateManager;
```
```suggestion
        [SerializeField] private UIStateManager stateManager;

        private UIStateEvent _onStateChange;
```

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateCaller.cs:
```diff
@@ -0,0 +1,18 @@
+
+        private void Start() => stateManager = GetComponent<UIStateManager>();
+
+        public void ChangeState(StateChange stateChange)
+        {
+            onStateChange?.Invoke(stateChange);
+            stateManager.OnStateChange(stateChange.states);
+        }
```
Missing summary

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,39 @@
+using System;
+using UI.Canvas.Introduction;
+using UI.Canvas.PhotoBookSystem;
+using UnityEngine;
```
```suggestion
using System;
using UnityEngine;

using UI.Canvas.Introduction;
using UI.Canvas.PhotoBookSystem;
```
[Code convention](https://github.com/Team-Swamp/CaptureGraafsmeer/wiki/Code-convention#importing-namespaces)

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace UI.Canvas.StateMachine
+{
+    public class UIStateManager : MonoBehaviour
```
```suggestion
    public sealed class UIStateManager : MonoBehaviour
```

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,39 @@
+{
+    public class UIStateManager : MonoBehaviour
+    {
+        private IntroductionPopup _introductionPopup;
+        private PhotoBook _photoBook;
+        private void Awake()
+        {
+            _introductionPopup = FindObjectOfType<IntroductionPopup>();
+            _photoBook = FindObjectOfType<PhotoBook>();
+        }
```
Zelf denk ik dat dit niet nodig is, er zijn maar 5 UI states en die hebben altijd het zelfde object. Hier kan makkelijk een standaard refrence gemaakt worden.

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,39 @@
+            _photoBook = FindObjectOfType<PhotoBook>();
+        }
+
+        public void OnStateChange(UIStates newState)
+        {
+            switch (newState)
+            {
+                case UIStates.MAIN:
+                    break;
+                case UIStates.INTRODUCTION:
+                    _introductionPopup.Open();
+                    break;
+                case UIStates.PHOTOBOOK:
+                    _photoBook.gameObject.SetActive(_photoBook.gameObject);
+                    break;
+                case UIStates.CAMERA_PANEL:
+                    break;
+                case UIStates.TAKING_PICTURE:
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(newState), newState, null);
+            }
+        }
```
```suggestion
        /// <summary>
        /// Fill in
        /// </summary>
        /// <param name="targetState">Parameter value to pass.</param>
        public void OnStateChange(UIStates targetState)
        {
            switch (targetState)
            {
                case UIStates.MAIN:
                    break;
                case UIStates.INTRODUCTION:
                    _introductionPopup.Open();
                    break;
                case UIStates.PHOTOBOOK:
                    _photoBook.gameObject.SetActive(_photoBook.gameObject);
                    break;
                case UIStates.CAMERA_PANEL:
                    break;
                case UIStates.TAKING_PICTURE:
                    break;
                default:
                    throw new ArgumentOutOfRangeException(nameof(newState), newState, null);
            }
        }
```

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStates.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public UIStates states;
+    }
+    public enum UIStates
+    {
+        MAIN,
+        INTRODUCTION,
+        PHOTOBOOK,
+        CAMERA_PANEL,
+        TAKING_PICTURE
+    }
```
```suggestion
    public enum UIState
    {
        MAIN,
        INTRODUCTION,
        PHOTOBOOK,
        CAMERA_PANEL,
        TAKING_PICTURE
    }
```
Dit hoort zijn eigen class te zijn. Zoals [dit](https://github.com/Team-Swamp/CaptureGraafsmeer/blob/develop/Assets/Scripts/Framework/Enums/CoordinatesTransformType.cs).

### Vergeet de name change niet!

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStates.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace UI.Canvas.StateMachine
+{
+    public class StateChange : MonoBehaviour
```
```suggestion
    public sealed class StateChange : MonoBehaviour
```

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStates.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class StateChange : MonoBehaviour
+    {
+        public UIStates states;
```
Het ding dat zoiezo anders moet is dat de `s` weg moet, het is 1 state en niet meerdere.
```suggestion
        public UIState state;
```

Dit is alleen public van wege [deze line 15](Assets/Scripts/UI/Canvas/StateMachine/UIStateCaller.cs). Hier kan makkelijk een getter voor gemaakt worden.

```suggestion
        [SerializeField] private UIState state;
        
        public UIState State => state;
```
OF
```suggestion
        [field: SerializeField] public UIState State{ get; private set; }
```

bas-boop reviewed (changes requested):
Wel goed op weg.

Wildeaso1 committed:
Moved scripts onto Main Canvas

Wildeaso1 committed:
Seperated Enum class and StateChange class

Wildeaso1 committed:
Applied Code feedback

Wildeaso1 committed:
Saved scene

bas-boop commented on Assets/Scripts/UI/Canvas/PhotoTaking/PhotoTaker.cs:
```diff
@@ -60,7 +60,7 @@ private void ApplyCamera(RawImage targetImage)
             if (CameraTexture == null)
                 FindCamera();
-            
+            Debug.Log("hello world");
```
```suggestion
            
```
Debug moet weg.

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIState.cs:
Staat in de verkeerde folder. `Assets/Scripts/Framwork/Enum`

bas-boop replied:
Vergeet niet om de namespace dan correct nog te hebben.

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateCaller.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        [SerializeField] private UIStateManager stateManager;
+
+        private UIStateEvent onStateChange;
```
```suggestion
        private UIStateEvent _onStateChange;
```
[code convention](https://github.com/Team-Swamp/CaptureGraafsmeer/wiki/Code-convention#variables)

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateEvent.cs:
Je mist de summary wat ik had gezegd in de vorige feedback.

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,56 @@
+using System;
+using Framework.PhoneCamera;
+using UnityEngine;
+
+using UI.Canvas.Introduction;
+using UI.Canvas.PhotoBookSystem;
+using UI.Canvas.PhotoTaking;
```
```suggestion
using System;
using UnityEngine;

using Framework.PhoneCamera;
using UI.Canvas.Introduction;
using UI.Canvas.PhotoBookSystem;
using UI.Canvas.PhotoTaking;
```
Standaard namespaces en onze moeten gesplits zijn.

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,56 @@
+        /// and whether or not you're allowed to move from a certain state to another one.
+        /// </summary>
+        /// <param name="newState">This is the state that is passed in.</param>
+        /// <exception cref="ArgumentOutOfRangeException">This error will be called when the state is unable</exception>
```
"when the state is ubable"

wsoijer replied:
> "when the state is ubable"

Er staat unable, moet unusable zijn

bas-boop replied:
Update het naar unusable in volgende commit.

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,56 @@
+                    break;
+                case UIState.CAMERA_PANEL when currentState == UIState.MAIN:
+                    cameraPanel.SetActive(true);
+                    /*takingPicture.SetActive(false);*/
```
```suggestion
```

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,56 @@
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(newState), newState, null);
+            }
+            currentState = newState;
```
```suggestion
            }
            
            currentState = newState;
```
[Styling convention](https://github.com/Team-Swamp/CaptureGraafsmeer/wiki/Code-convention#bracket-placement)

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,56 @@
+                    cameraPanel.SetActive(false);
+                    takingPicture.SetActive(false);
+                    break;
+                case UIState.INTRODUCTION when currentState == UIState.MAIN:
+                    introductionPopup.Open();
+                    break;
+                case UIState.PHOTOBOOK when currentState == UIState.MAIN:
+                    photoBook.gameObject.SetActive(photoBook.gameObject);
+                    break;
+                case UIState.CAMERA_PANEL when currentState == UIState.MAIN:
+                    cameraPanel.SetActive(true);
+                    /*takingPicture.SetActive(false);*/
+                    break;
+                case UIState.TAKING_PICTURE when currentState == UIState.CAMERA_PANEL:
```
```suggestion
                case UIState.INTRODUCTION
                when currentState == UIState.MAIN:
                    introductionPopup.Open();
                    break;
                case UIState.PHOTOBOOK
                when currentState == UIState.MAIN:
                    photoBook.gameObject.SetActive(photoBook.gameObject);
                    break;
                case UIState.CAMERA_PANEL
                when currentState == UIState.MAIN:
                    cameraPanel.SetActive(true);
                    /*takingPicture.SetActive(false);*/
                    break;
                case UIState.TAKING_PICTURE
                when currentState == UIState.CAMERA_PANEL:
```
IMO is dit meer leesbaarder. Lijkt op de [regel van meerdere conditions van een if statement](https://github.com/Team-Swamp/CaptureGraafsmeer/wiki/Code-convention#if-statements).

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,56 @@
+        /// </summary>
+        /// <param name="newState">This is the state that is passed in.</param>
+        /// <exception cref="ArgumentOutOfRangeException">This error will be called when the state is unable</exception>
+        public void SwitchState(UIState newState)
+        {
+            switch (newState)
```
```suggestion
        public void SwitchState(UIState targetState)
        {
            switch (targetState)
```
Dit stond ook in een vorige feedback punt. Dit zegt dat we switchen naar een andere target state. IPV dat we switchen naar een nieuwe aangemaakte state.

bas-boop commented on Assets/Scripts/UI/Canvas/StateMachine/UIStateManager.cs:
```diff
@@ -0,0 +1,56 @@
+{
+    public sealed class UIStateManager : MonoBehaviour
+    {
+        [Header("References")]
+        [SerializeField] private IntroductionPopup introductionPopup;
+        [SerializeField] private PhotoBook photoBook;
+        [SerializeField] private GameObject cameraPanel;
+        [SerializeField] private GameObject takingPicture;
+        
+        [Header("Current UI State")]
+        [SerializeField] private UIState currentState;
```
```suggestion
        [SerializeField] private UIState currentState;

        [Header("References")]
        [SerializeField] private IntroductionPopup introductionPopup;
        [SerializeField] private PhotoBook photoBook;
        [SerializeField] private GameObject cameraPanel;
        [SerializeField] private GameObject takingPicture;
```
Het is nu "Current UI state: current state", dit vind ik niet echt mooi, dus verplaats het naar boven zonder header. Waardoor alleen de refrences een header hebben.

OF

Bedenk een betere naam voor de 2de header.

bas-boop reviewed (changes requested):
Zet de kers op de taart.

Wildeaso1 committed:
Applied code feedback

Turns out the custom Unity event that I had created was unnecessary and has been deleted. The UIStateCaller has been changed to reflect this change.

Wildeaso1 committed:
Changed name

bas-boop reviewed (changes requested):
# Heb wat dingen unresolved van vorige reviews.

Wildeaso1 committed:
changed name

bas-boop reviewed (approved):
*paarse vroom vroom* van William
